os/arch/arm/src/amebasmart/amebasmart_i2c.c

```diff
@@ -1226,3 +1249,149 @@ int up_i2creset(FAR struct i2c_dev_s *dev)
 	return 0;
 }
 #endif							/* CONFIG_I2C_RESET */
+
+/****************************************************************************
+ * Name: rtk_i2c_suspend/resume
+ *
+ * Description:
+ *   Suspend or resume i2c peripherals for/from sleep modes.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_PM
+static uint32_t rtk_i2c_suspend(uint32_t expected_idle_time, void *param)
+{
+	(void)expected_idle_time;
+	(void)param;
+
+	/* Need to destroy semaphore and release i2c object to prevent memory leak */
+#ifdef CONFIG_AMEBASMART_I2C1
+	up_i2cuninitialize(&amebasmart_i2c1_priv);
+#endif
+#ifdef CONFIG_AMEBASMART_I2C2
+	up_i2cuninitialize(&amebasmart_i2c2_priv);
+#endif
+	return 1;
+}
+
+static uint32_t rtk_i2c_resume(uint32_t expected_idle_time, void *param)
+{
+	(void)expected_idle_time;
+	(void)param;
+	/* For PG Sleep, I2C 1/2 HW will be lost power, thus a reinitialization is required here 
+	I2C0 is under SYSON block, so it will not require to be reinitialized*/
+#ifdef CONFIG_AMEBASMART_I2C1
+	(void) up_i2cinitialize(1);
+#endif
+#ifdef CONFIG_AMEBASMART_I2C1
+	(void) up_i2cinitialize(2);
+#endif
+	return 1;
+}
+#endif
+
+/****************************************************************************
+ * Name: amebasmart_i2c_pmnotify
+ *
+ * Description:
+ *   Notify the driver of new power state. This callback is  called after
+ *   all drivers have had the opportunity to prepare for the new power state.
+ *
+ * Input Parameters:
+ *
+ *    cb - Returned to the driver. The driver version of the callback
+ *         structure may include additional, driver-specific state data at
+ *         the end of the structure.
+ *
+ *    pmstate - Identifies the new PM state
+ *
+ * Returned Value:
+ *   None - The driver already agreed to transition to the low power
+ *   consumption state when when it returned OK to the prepare() call.
+ *
+ *
+ ****************************************************************************/
+#ifdef CONFIG_PM
+static void amebasmart_i2c_pmnotify(FAR struct pm_callback_s *cb, int domain,
+                                   enum pm_state_e pmstate)
+{
+	switch (pmstate)
+	{
+		case PM_NORMAL:
+			pmu_unregister_sleep_callback(PMU_I2C_DEVICE);
```

Comment:
This means the pm state should be normal at wakeup? I heard it could be the standby when LCD is used. For example, lcd should not be turned on but something should be working (main core should be wakeup without lcd on). In this case, is this right? @ritesh55555 please check this.

Reply:
1. Yes , for now , whenever board wakes up with an interrupt, pm state goes to NORMAL .
2. When LCD is  used, according the design given (lcd and pm) , pm should go to stand by state . So if we do that , we need to change this part and also in some other files also.

Reply:
PM_NORMAL can be set in several situations.

This may not be the case when the main core wakes up.

I think it's a better way to restore this function directly from the pmu code in BSP than this.

The noti callback of this PM is not a feature for the BSP stage.

Reply:
Hi all, I have pushed a commit to move the pre/post sleep callbacks to being done only during initialization stage.